src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/utils/CardinalityEstimation.java

```diff
@@ -0,0 +1,291 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import org.apache.jena.sparql.core.Var;
+import se.liu.ida.hefquin.engine.federation.*;
+import se.liu.ida.hefquin.engine.federation.access.*;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.BGPRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.SPARQLRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.TPFRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.TriplePatternRequestImpl;
+import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.queryplan.LogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.*;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalOpRequest;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithNullaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.utils.ExpectedVariablesUtils;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryProcContext;
+
+import java.util.*;
+
+public class CardinalityEstimation {
+    protected final CardinalitiesCache cardinalitiesCache = new CardinalitiesCache();
+    protected final VarSpecificCardinalitiesCache varSpecificCardinalitiesCache = new VarSpecificCardinalitiesCache();
+    protected final QueryProcContext ctxt;
+
+    public CardinalityEstimation(QueryProcContext ctxt) {
+        assert ctxt != null;
+        this.ctxt = ctxt;
+    }
+
+    public int getCardinalityEstimationOfLeafNode( final PhysicalPlan pp ) throws QueryOptimizationException, FederationAccessException {
```

Comment:
Please read my previous comment about the exceptions again. I think that this method (and the other public methods of this class) should not throw a FederationAccessException but they may only throw a QueryOptimizationException.

Hence, in the implementation of this method, the function calls that may throw a FederationAccessException should be wrapped in a try-catch block where the catch branch would then throw a new QueryOptimizationException as described in my previous comment.